rclcpp/include/rclcpp/strategies/msg_pool_memory_strategy.hpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+#define RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+
+#include <rclcpp/msg_memory_strategy.hpp>
+#include <rclcpp/macros.hpp>
+
+namespace rclcpp
+{
+namespace msg_pool_memory_strategy
+{
+
+template<typename MessageT, size_t size, typename std::enable_if<rosidl_generator_traits::has_fixed_size<MessageT>::value>::type * = nullptr>
+class MessagePoolMemoryStrategy
+  : public msg_memory_strategy::MessageMemoryStrategy<MessageT>
+{
+public:
+  RCLCPP_MAKE_SHARED_DEFINITIONS(MessagePoolMemoryStrategy);
+  MessagePoolMemoryStrategy()
+  : seq_(0)
+  {
+    for (size_t i = 0; i < size; ++i) {
+      msg_ptrs_[i] = std::make_shared<MessageT>(msg_pool_[i]);
```

Comment:
I don't think `make_shared` does what you think (or what I think you think based on this line):

http://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared
> Constructs an object of type T and wraps it in a std::shared_ptr using args as the parameter list for the constructor of T.

So you're actually just passing the instance in `msg_pool_` to the *copy* constructor of a newly allocated version in `msg_ptrs_`. So it's making a copy and so I don't think you need `msg_pool_` at all here.